File: MCTS_BlokusDuo_cpp/Player.cpp
//
// Created by joaovieira on 12/24/19.
//

#include "Player.h"

uint8_t Player::idCount = 0;

Player::Player() {
    playerId = ++idCount;
    remainingGamePieces = GamePiece::getInitialSetOfGamePieces();
}

uint8_t Player::getPlayerId() {
    return playerId;
}

Player::~Player() {
    delete remainingGamePieces;
}

std::list<std::shared_ptr<std::vector<std::shared_ptr<GamePiece>>>> *Player::getRemainingGamePieces() {
    return remainingGamePieces;
}

Player::Player(Player *player, std::shared_ptr<GamePiece> removePiece) {
    playerId = player->playerId;

    remainingGamePieces = new std::list<std::shared_ptr<std::vector<std::shared_ptr<GamePiece>>>>();
    for (auto &gamePiece : *player->remainingGamePieces)
        if ((*gamePiece)[0]->getCodeName() != removePiece->getCodeName())
            remainingGamePieces->push_back(gamePiece);

    quited = player->quited;
}

Player::Player(Player *player) {
    playerId = player->playerId;

    //remainingGamePieces = player->remainingGamePieces;
    remainingGamePieces = new std::list<std::shared_ptr<std::vector<std::shared_ptr<GamePiece>>>>();
    for (auto &gamePiece : *player->remainingGamePieces)
        remainingGamePieces->push_back(gamePiece);

    quited = player->quited;
}

bool Player::getQuited() {
    return quited;
}

int Player::getScore() {
    int score = 0;

    for (auto &gamePiece : *remainingGamePieces)
        score += (*gamePiece)[0]->getSquares()->size();

    return score;
}

void Player::setQuited(bool quited) {
    this->quited = quited;
}

std::string Player::toString() {
    std::string text;

    text += "Player " + std::to_string(playerId) + ": { ";
    text += "quited: ";
    if (quited) text += "true; "; else text += "false; ";
    text += "score: " + std::to_string(getScore()) + "; ";
    text += "remaining pieces: ";
    for (auto gamePiece : *remainingGamePieces) {
        text += (*gamePiece)[0]->getCodeName();
        text += " ";
    }

    text += "}";

    return text;
}

File: MCTS_BlokusDuo_java/src/com/joaomiguelvieira/Player.java
package com.joaomiguelvieira;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;

public class Player {
    private static byte idCount = 0;

    private byte playerId;
    private boolean quited = false;
    private ArrayList<ArrayList<GamePiece>> remainingGamePieces;

    public Player() {
        playerId = ++idCount;
        remainingGamePieces = GamePiece.getInitialSetOfGamePieces();
    }

    public Player(Player player, GamePiece removePiece) {
        playerId = player.playerId;

        remainingGamePieces = new ArrayList<>();
        for (ArrayList<GamePiece> gamePiece : player.remainingGamePieces)
            if (gamePiece.get(0).getCodeName() != removePiece.getCodeName())
                remainingGamePieces.add(gamePiece);

        quited = player.quited;
    }

    public Player(Player player) {
        playerId = player.playerId;

        remainingGamePieces = new ArrayList<>();
        remainingGamePieces.addAll(player.remainingGamePieces);

        quited = player.quited;
    }

    public byte getPlayerId() {
        return playerId;
    }

    public ArrayList<ArrayList<GamePiece>> getRemainingGamePieces() {
        return remainingGamePieces;
    }

    public boolean getQuited() {
        return quited;
    }

    public int getScore() {
        int score = 0;

        for (ArrayList<GamePiece> gamePiece : remainingGamePieces)
            score += gamePiece.get(0).getSquares().length;

        return score;
    }

    public void setQuited(boolean quited) {
        this.quited = quited;
    }

    @Override
    public String toString() {
        StringBuilder text = new StringBuilder();

        text.append("Player ").append(playerId).append(": { ");

        text.append("quited: ").append(quited ? "true; " : "false; ");
        text.append("score: ").append(getScore()).append("; ");
        text.append("remaining pieces: ");
        for (ArrayList<GamePiece> gamePiece : remainingGamePieces)
            text.append(gamePiece.get(0).getCodeName()).append(" ");

        text.append("}");

        return text.toString();
    }
}
